### ISSUE TITLE
Shouldn't need to add generic builder factories for serialization by hand

### ISSUE THREAD

**Issue by davidmorgan** [google/built_value.dart#124]
See test

https://github.com/google/built_value.dart/blob/master/end_to_end_test/test/generics_serializer_test.dart 

Currently, serialization with generics can require manually adding builder factories to "serializers". This should be done for you.

**Comment by BrennanGambling** [comment]
Making some good progress on this issue. Is it fine if I modify the tests in generic_serializer_tests to reflect the fact that a missing builder (not really missing anymore) for a generic should no longer be a problem?

**Comment by BrennanGambling** [comment]
Also should builders be generated for a generic parameter that itself has a generic parameter. For example the CollectionGenericValue<T> could have T set as a BuiltList<T>. So a builder for CollectionGenericValue<BuiltList> will be made but not for all of the possible BuiltList<T> like BuiltList<String> or BuiltList<int>.
The biggest issue with this is a generic type parameter could be set to a type with a generic type parameter that could be also has a generic type parameter. Kind of like this.
BuiltList<BuiltList<BuiltList<BuiltList<...
Should this be implemented and what would the maximum depth be.

**Comment by davidmorgan** [comment]
(I fixed your comment so the angular brackets are visible--by default they disappear).

Sure, go ahead and modify the tests.

There isn't a way to generate all the builders needed where there's an unbound `T` at the top level, since we don't know what `T` could be. We can only generate builders for cases where a field is used that fills in the type. For example if there is another class `Foo` that has a `CollectionGenericValue<int>` then we know we'll need `BuiltList<int>`.

**Comment by BrennanGambling** [comment]
Yeah I was thinking just of creating builders with `T` as all of the currently serializable types (the ones added to the base `Serializers` plus the ones created in the annotation. And then the same for a Bound `T` but only when that type is a subtype of the the bound `T`.
The biggest issue is it would add a lot of builders. Like if a class had 2 unbound generic parameters just the serializable types in the base `Serializers` would result in 17*17=289 builders a single class.
I'm going to try to use the analyzers AST library to look for cases where an object is being serialized or deserialized with a specified FullType, and just add builders for those.

**Comment by davidmorgan** [comment]
Right, it's not feasible to generate for everything that _could_ be used--and in particular people could use any of their user defined types. So we can stick with just things we know will be used.

**Comment by davidmorgan** [comment]
It's too expensive to process the entire source; what we need is to look at all serializable types and their fields. This won't catch everything but it will catch most cases.

For example, if a serializable class has a field of type `MyValue<Foo>` and `MyValue<T>` has a field `BuiltList<T>`, then we should generate a builder for `BuiltList<Foo>`.

**Comment by swch01** [comment]
I believe this belongs here. If I create a generic built value that contains two BuiltMap fields, one of which uses a fixed rather than generic type, the code generator fails to add a builder factory for that BuiltMap. (When I add the factory by hand, serialization works.)

Class:
```
abstract class Foo<K, V> implements Built<Foo<K, V>, FooBuilder<K, V>> {
  factory Foo([void Function(FooBuilder<K, V>) updates]) = _$Foo<K, V>;
  Foo._();
  static Serializer<Foo> get serializer => _$fooSerializer;

  BuiltMap<K, V> get field1;
  BuiltMap<K, bool> get field2;
}
```

Serializers:
```
@SerializersFor([Foo])
final Serializers serializers = (_$serializers.toBuilder()).build();

```

Generated code:
```
Serializers _$serializers =
    (new Serializers().toBuilder()..add(Foo.serializer)).build();
```

Error:
```
Bad state: No builder factory for BuiltMap<Object, bool>. Fix by adding one, see SerializersBuilder.addBuilderFactory.
```

**Comment by davidmorgan** [comment]
There are some cases where codegen could be smarter and add factories for you--which is what this bug is about--not sure if the case you list is one of them, it depends on whether codegen can figure out that you might use `Foo<Object, ...>` somewhere.
